plaidcloud/rpc/functions.py

```diff
@@ -8,24 +8,59 @@
 Note: more basic ones can be found in toolz
 """
 
-from __future__ import absolute_import
 import multiprocessing
 import sys
 import traceback
 import re
-from six.moves import reduce
+import asyncio
+from collections.abc import Iterable, Collection
+from typing import TypeVar, Callable, Union, Any
+from functools import reduce
 
 from toolz.dicttoolz import merge_with
-from functools import reduce
 
 __author__ = 'Adams Tower'
 __credits__ = ['Adams Tower', 'Paul Morel']
 __maintainer__ = 'Adams Tower'
-__copyright__ = '© Copyright 2011-2018 Tartan Solutions, Inc.'
+__copyright__ = '© Copyright 2011-2021 Tartan Solutions, Inc.'
 __license__ = 'Apache 2.0'
 
+T = TypeVar('T')
+
+
+def async_memoize(afunc):
+    """Decorator to memoize an async function."""
+    #TODO: switch out for a library rather than own-rolled: https://github.com/aio-libs/async-lru"""
+    #TODO: if we don't, figure out how to type hint this
+    cache: dict = {}
+    async def memoized_afunc(*args, **kwargs):
+        key = (args, frozenset(sorted(kwargs.items())))
+        if key in cache:
+            return cache[key]
+        result = await afunc(*args, **kwargs)
+        cache[key] = result
+        return result
+    return memoized_afunc
+
 
-def try_except(success, failure):
+async def gather_with_sem(*futures, concurrent_tasks=3):
```

Comment:
Looks ok, but I don't like the name much. 
I can't think of much better, perhaps call it `gather_with_semaphore` fq
or perhaps doesn't really matter